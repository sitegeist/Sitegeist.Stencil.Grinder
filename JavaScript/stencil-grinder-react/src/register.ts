import { ComponentType } from "react";

/**
 * This file is part of the Sitegeist.Stencil.Grinder package
 *
 * (c) 2019
 * Wilhelm Behncke <behncke@sitegeist.de>
 *
 * This package is Open Source Software. For the full copyright and license
 * information, please view the LICENSE file which was distributed with this
 * source code.
 */

/**
 * Central registry for all components known to Grinder
 */
const registry: Map<string, ComponentType<any>> = new Map();

/**
 * @private
 * @param string key
 * @return ComponentType<any>
 */
export function get(key: string): ComponentType<any> {
	if (!registry.has(key)) {
		throw new Error(`[Grinder]: No component has been registered for key "${key}".`);
	}

	return registry.get(key) as ComponentType<any>;
}

/**
 * @public
 * @param string key
 * @param ComponentType implementation?
 * @return ComponentType<any> | (implementation: ComponentType<any>) => ComponentType<any>
 */
export default function register(key: string, implementation?: ComponentType<any>) {
	if (implementation) {
		registry.set(key, implementation);
		return implementation;
	}

	return function registerDecorator(implementation: ComponentType<any>) {
		registry.set(key, implementation);
		return implementation;
	}
}
