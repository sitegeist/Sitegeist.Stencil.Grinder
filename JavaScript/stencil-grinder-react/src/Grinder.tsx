import React from "react";
import { GrinderResponse } from "@sitegeist/stencil-grinder-shared";

import { get } from "./register";

/**
 * This file is part of the Sitegeist.Stencil.Grinder package
 *
 * (c) 2019
 * Wilhelm Behncke <behncke@sitegeist.de>
 *
 * This package is Open Source Software. For the full copyright and license
 * information, please view the LICENSE file which was distributed with this
 * source code.
 */

type GrinderData = GrinderResponse | { key: any, [key: string]: any } | object | number | string | null;

interface GrinderProps {
	data: GrinderData | GrinderData[]
}

/**
 * @public
 * @param GrinderProps props
 * @return JSX.Element[] | undefined
 */
export default function Grinder(props: GrinderProps): JSX.Element | null {
	if (Array.isArray(props.data)) {
		return (
			<>
				{props.data.map((item, index) => {
					if (item !== null && typeof item === 'object') {
						return (
							<Grinder key={'key' in item ? item.key : index} data={item} />
						);
					}

					return item;
				})}
			</>
		);
	}

	if (typeof props.data === 'object' && props.data !== null && 'type' in props.data) {
		switch (props.data.type) {
			case 'Sitegeist.Stencil.Grinder/v1/INDEX':
				throw new Error(`[Grinder]: Cannot render index. The index is not supposed to be rendered.`);

			case 'Sitegeist.Stencil.Grinder/v1/COMMAND':
				throw new Error(`[Grinder]: Cannot render command "${props.data.payload.directive}". Please make sure to handle commands before rendering.`);

			case 'Sitegeist.Stencil.Grinder/v1/DOCUMENT':
				return (
					<Grinder data={props.data.payload}/>
				);

			case 'Sitegeist.Stencil.Grinder/v1/NODE': {
				const Component = get(props.data.payload.component);

				if (Component) {
					return (<Component {...props.data.payload.props} />);
				}

				throw new Error(`[Grinder]: Could not find component: ${props.data.payload.component}`);
			}

			default: break;
		}
	}

	if (typeof props.data === 'string') {
		return (<>{props.data}</>);
	}

	if (typeof props.data === 'number') {
		return (<>props.data</>);
	}

	if (props.data === null) {
		return null;
	}

	throw new Error(`[Grinder]: Could not unravel data structure: ${JSON.stringify(props.data)}.`);
}
