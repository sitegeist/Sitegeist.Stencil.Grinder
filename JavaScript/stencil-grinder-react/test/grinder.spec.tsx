import React from "react";
import renderer from "react-test-renderer";
import register from "../src/register";
import Grinder from "../src/Grinder";

describe(`Grinder`, () => {
	beforeAll(() => {
		register('Site', props => (<div><Grinder data={props.children}/></div>));
		register('Header', () => (<header>HEADER</header>));
		register('Body', (props: { content: string }) => (<div>{props.content}</div>));
		register('Footer', () => (<footer>FOOTER</footer>));
	});

	it(`Should render a node`, () => {
		const rendered = renderer.create(
			<Grinder
				data={{
					type: 'Sitegeist.Stencil.Grinder/v1/NODE',
					payload: {
						component: 'Body',
						props: {
							content: 'Foo'
						}
					}
				}}
			/>
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`Should render an array of nodes`, () => {
		const rendered = renderer.create(
			<Grinder
				data={[
					{
						type: 'Sitegeist.Stencil.Grinder/v1/NODE',
						payload: {
							component: 'Body',
							props: {
								content: 'Foo'
							}
						}
					},
					{
						type: 'Sitegeist.Stencil.Grinder/v1/NODE',
						payload: {
							component: 'Body',
							props: {
								content: 'Bar'
							}
						}
					}
				]}
			/>
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`Should render a document`, () => {
		const rendered = renderer.create(
			<Grinder
				data={{
					type: 'Sitegeist.Stencil.Grinder/v1/DOCUMENT',
					payload: {
						type: 'Sitegeist.Stencil.Grinder/v1/NODE',
						payload: {
							component: 'Site',
							props: {
								children: [
									{
										type: 'Sitegeist.Stencil.Grinder/v1/NODE',
										payload: {
											component: 'Header'
										}
									},
									{
										type: 'Sitegeist.Stencil.Grinder/v1/NODE',
										payload: {
											component: 'Body',
											props: {
												content: 'Foo'
											}
										}
									},
									{
										type: 'Sitegeist.Stencil.Grinder/v1/NODE',
										payload: {
											component: 'Footer'
										}
									}
								]
							}
						}
					}
				}}
			/>
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through strings`, () => {
		const rendered = renderer.create(
			<Grinder data={'Foo'} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through arrays of strings`, () => {
		const rendered = renderer.create(
			<Grinder data={['Foo', 'Bar']} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through numbers`, () => {
		const rendered = renderer.create(
			<Grinder data={12} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through arrays of numbers`, () => {
		const rendered = renderer.create(
			<Grinder data={[12, 34]} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through null`, () => {
		const rendered = renderer.create(
			<Grinder data={null} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should pass through arrays of null`, () => {
		const rendered = renderer.create(
			<Grinder data={[null, null, null]} />
		);

		expect(rendered).toMatchSnapshot();
	});

	it(`should throw, if attempted to render a command`, () => {
		expect(() => {
			renderer.create(
				<Grinder
					data={{
						type: 'Sitegeist.Stencil.Grinder/v1/COMMAND',
						payload: {
							directive: 'REDIRECT',
							options: {
								location: '/somewhere'
							}
						}
					}}
				/>
			);
		}).toThrowErrorMatchingSnapshot();
	});

	it(`should throw, if attempted to render an index`, () => {
		expect(() => {
			renderer.create(
				<Grinder
					data={{
						type: 'Sitegeist.Stencil.Grinder/v1/INDEX',
						payload: []
					}}
				/>
			);
		}).toThrowErrorMatchingSnapshot();
	});

	it(`should throw, if attempted to render anything else`, () => {
		expect(() => {
			renderer.create(
				<Grinder data={{foo: 'bar'}}/>
			);
		}).toThrowErrorMatchingSnapshot();

		expect(() => {
			renderer.create(
				<Grinder data={[{foo: 'bar'}]}/>
			);
		}).toThrowErrorMatchingSnapshot();
	});
})
