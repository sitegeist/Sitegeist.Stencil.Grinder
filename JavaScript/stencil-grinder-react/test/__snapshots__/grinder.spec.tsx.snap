// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grinder Should render a document 1`] = `
<div>
  <header>
    HEADER
  </header>
  <div>
    Foo
  </div>
  <footer>
    FOOTER
  </footer>
</div>
`;

exports[`Grinder Should render a node 1`] = `
<div>
  Foo
</div>
`;

exports[`Grinder Should render an array of nodes 1`] = `
Array [
  <div>
    Foo
  </div>,
  <div>
    Bar
  </div>,
]
`;

exports[`Grinder should pass through arrays of null 1`] = `null`;

exports[`Grinder should pass through arrays of numbers 1`] = `
Array [
  "12",
  "34",
]
`;

exports[`Grinder should pass through arrays of strings 1`] = `
Array [
  "Foo",
  "Bar",
]
`;

exports[`Grinder should pass through null 1`] = `null`;

exports[`Grinder should pass through numbers 1`] = `"props.data"`;

exports[`Grinder should pass through strings 1`] = `"Foo"`;

exports[`Grinder should throw, if attempted to render a command 1`] = `"[Grinder]: Cannot render command \\"REDIRECT\\". Please make sure to handle commands before rendering."`;

exports[`Grinder should throw, if attempted to render an index 1`] = `"[Grinder]: Cannot render index. The index is not supposed to be rendered."`;

exports[`Grinder should throw, if attempted to render anything else 1`] = `"[Grinder]: Could not unravel data structure: {\\"foo\\":\\"bar\\"}."`;

exports[`Grinder should throw, if attempted to render anything else 2`] = `"[Grinder]: Could not unravel data structure: {\\"foo\\":\\"bar\\"}."`;
